### PR TITLE
HIP Backend Support

### PR DESCRIPTION
**Pull request by rcarson3**
Draft proposal to add in HIP backend support to ExaCMech

### CODE REVIEW

**Comment by rcarson3** [comment]
All of the examples should now pass when compiled with a HIP compiler. I'll work on porting the miniapp over next.

**Comment by rcarson3** [comment]
As of https://github.com/LLNL/ExaCMech/pull/5/commits/0cfb7bfb2cc58a99c8ea120e42d257711b6e8c40 the miniapp should now work with HIP / ROCm 4.2 as built on Spock and using RAJA v0.14. I will note that a few additional compiler options were  needed to get to this point. One of them is that do need to make use of C++14 in-order to properly compile everything.

**Comment by rcarson3** [comment]
Also, apparently if we do a HIP build we will no longer be able to do an openmp+device build of things based on the below similar issue...

https://github.com/LLNL/RAJA/issues/976

**Comment by rcarson3** [comment]
In order to drastically reduce the amount of repeat code we have for each forall implementation, I'm going to look into replacing all these `forall` loops with either the `RAJA::launch` or `RAJA::expt::dynamic_forall` feature (https://github.com/LLNL/RAJA/pull/1280). This will push the minimum RAJA version up to `v2022.10.x`, but I don't see an issue with that if that means we'll have less code to worry about. Although, if we do move towards the launch syntax that might cause us to lose out on the ability to use `OpenMP` with the host and GPU codes. However, I don't necessarily see that as much of an issue now given my experience over the years running on GPU systems. Since, you almost always just run on the GPU or CPU. Outside of benchmarking, I can't remember the last time I used OpenMP on a GPU machine...

It might take just a bit to get this branch refactored to that as I need to keep this branch stable at least until I run some large runs on Frontier.

Also before merging this, I need to simplify all the `CUDA` vs `HIP` logic in terms of execution strategy and the cuda portability header and just have things referred to as `GPU`.

File: README.md [hunk 1]
```diff
@@ -56,7 +56,7 @@ The develop branch is the main development branch. Changes to develop are by pul
 MINIAPP
 ======
 
-A miniapp is provided which provides a guide for how the library might be used in a real library. It provides several compute kernels that contain CPU, OpenMP, and CUDA code through the use of RAJA. A bash script is provided for a CORAL machine that runs a few different examples using the CPU, OpenMP, and CUDA for realistic size data.
+A miniapp is provided which provides a guide for how the library might be used in a real library. It provides several compute kernels that contain CPU, OpenMP, and CUDA/HIP code through the use of RAJA. A bash script is provided for a CORAL machine that runs a few different examples using the CPU, OpenMP, and CUD/HIP for realistic size data.
```
Comment (nrbarton) [inline README.md]: Maybe "used in a real application" would be more accurate than "used in a real library"

File: test/test_updst.cxx [hunk 2]
```diff
@@ -37,8 +37,12 @@ TEST(ecmech, updst_a)
 
 #if KIN_TYPE
    matModelEvptn_FCC_B* mmodel = new matModelEvptn_FCC_B();
+   constexpr int nH = matModelEvptn_FCC_B::nH;
```
Comment (nrbarton) [inline test/test_updst.cxx]: I am going to show how rusty I am with C++ by asking this -- to help prevent against future cut-and-paste errors, can you get nH and nslip from the mmodel instance rather than from the class? Same for the FCC_A case just below.

File: src/ecmech/ECMech_evptnWrap.h [hunk 3]
```diff
@@ -357,8 +357,37 @@ namespace ecmech {
                });
                      break;
 #endif
-                  case ECM_EXEC_STRAT_CPU:
-                  default: // fall through to CPU if other options are not available
+#if defined(RAJA_ENABLE_HIP)
+                  case ECM_EXEC_STRAT_GPU :
```
Comment (nrbarton) [inline src/ecmech/ECMech_evptnWrap.h]: It looks like the code is only set up for either RAJA_ENABLE_CUDA or RAJA_ENABLE_HIP to be defined. It makes sense that you would only have one or the other ... but is there something to make that even more transparent? Would it be better if this were an #elif rather than a #if?

File: src/ecmech/ECMech_gpu_portability.h [hunk 4]
```diff
@@ -54,14 +46,18 @@ extern void CUDART_Check(const cudaError_t err, const char *file, const char *fu
...
-#define __cuda_device_only__
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ > 0))  || defined(__HIP_DEVICE_COMPILE__)
+#define __ecmech_device_only__
 #else
-#define __cuda_host_only__
+#define __ecmech_host_only__
 #endif
 
-#endif // __ECMECH_CUDA_PORTABILITY_H
+#endif // __ECMECH_GPU_PORTABILITY_H
```
Comment (nrbarton) [inline src/ecmech/ECMech_gpu_portability.h]: It looks like you still need to change __ECMECH_CUDA_PORTABILITY_H in lines 3 and 4 to __ECMECH_GPU_PORTABILITY_H

File: src/ecmech/ECMech_util.h [hunk 5]
```diff
@@ -1192,8 +1190,8 @@ namespace ecmech {
             mtanSD(jSvec, 5) = C(jSvec, 2) * sqr2;
          }
       }
-   } // mtan_conv_sd_svec
-
+   } // mtan_conv_sd_
```
Comment (nrbarton) [inline src/ecmech/ECMech_util.h]: Why the change here? -- It does not look like the function name changed.

File: miniapp/miniapp_util.h [hunk 6]
```diff
@@ -28,31 +32,76 @@ enum class ExecutionStrategy { CPU, CUDA, OPENMP };
...
+#elif defined(RAJA_ENABLE_HIP)
+      hipErrchk(hipFree(ptr));
+#endif
+      ptr = nullptr;
+    }
+  }
+#endif
 }; // namespace memoryManager
 
 // Copyright (c) 2016-19, Lawrence Livermore National Security, LLC.
```
Comment (nrbarton) [inline miniapp/miniapp_util.h]: Is all of this copyright boilerplate up to date and what you still want to have in there? Should the copyright dates get updated?

File: miniapp/orientation_evolution.cxx [hunk 7]
```diff
@@ -481,15 +521,42 @@ int main(int argc, char *argv[]){
...
+                  const double* stress = &(d_stress_array[i_qpts * ecmech::nsvec]);
                   cuda_sum += wts * stress[j];
                });
                stress_avg[j] = cuda_sum.get();
             }
          }
          break;
 #endif
-      } // switch ( class_device )
+#if defined(RAJA_ENABLE_HIP)
```
Comment (nrbarton) [inline miniapp/orientation_evolution.cxx]: Changing from #if to #elif might improve readability?

File: miniapp/retrieve_kernels.cxx [hunk 8]
```diff
@@ -141,12 +185,20 @@ void retrieve_data(ecmech::ExecutionStrategy accel, const int nqpts, const int n
...
 #if defined(RAJA_ENABLE_CUDA)
-      case ECM_EXEC_STRAT_CUDA:
+      case ECM_EXEC_STRAT_GPU:
       {
          retrieve_data_cuda(nqpts, nstatev, stress_svec_p_array, vol_ratio_array,
                             eng_int_array, state_vars_array, stress_array);
       }
       break;
+#endif
+#if defined(RAJA_ENABLE_HIP)
```
Comment (nrbarton) [inline miniapp/retrieve_kernels.cxx]: Change to #elif to improve readability?

File: miniapp/retrieve_kernels.cxx [hunk 9]
```diff
@@ -123,6 +123,50 @@ namespace {
...
+   void retrieve_data_hip(const int nqpts, const int nstatev,
+                          const double* stress_svec_p_array, const double* vol_ratio_array,
+                          const double* eng_int_array, double* state_vars_array,
+                          double* stress_array){
+      const int ind_int_eng = nstatev - ecmech::ne;
+      const int ind_vols = ind_int_eng - 1;
+
+      RAJA::RangeSegment default_range(0, nqpts);
+
+      RAJA::forall<RAJA::hip_exec<384> >(default_range, [ = ] RAJA_DEVICE(int i_qpts) {
```
Comment (nrbarton) [inline miniapp/retrieve_kernels.cxx]: This is a lot of code duplication for what appears to be a on-line difference in the RAJA::forall invocation. 

Is there some not-overly-painful way to template this to avoid the code duplication?

File: miniapp/setup_kernels.cxx [hunk 10]
```diff
@@ -383,6 +383,98 @@ namespace {
...
+
+      const int DIM = 3;
+      std::array<RAJA::idx_t, DIM> perm { { 2, 1, 0 } };
+      RAJA::Layout<DIM> layout = RAJA::make_permuted_layout({ { ecmech::ndim, ecmech::ndim, nqpts } }, perm);
+      RAJA::View<const double, RAJA::Layout<DIM, RAJA::Index_type, 0> > vgrad_view(vel_grad_array, layout);
+
+      // All of the below we could setup in one big RAJA loop/kernel
+      RAJA::RangeSegment default_range(0, nqpts);
+
+      RAJA::forall<RAJA::hip_exec<384> >(default_range, [ = ] RAJA_DEVICE(int i_qpts) {
```
Comment (nrbarton) [inline miniapp/setup_kernels.cxx]: his is a lot of code duplication for what appears to be a on-line difference in the RAJA::forall invocation.

Is there some not-overly-painful way to template this to avoid the code duplication?

File: src/ecmech/ECMech_util.h (hunk 5 above)
Comment (rcarson3) [reply src/ecmech/ECMech_util.h]: Must have accidentally deleted this I'll revert the change.

File: src/ecmech/ECMech_evptnWrap.h (hunk 3 above)
Comment (rcarson3) [reply src/ecmech/ECMech_evptnWrap.h]: The `#elif` would likely make more sense here. The old version is likely a by-product when I initially added things and both CUDA/HIP were supported at the same time. I'll update that.

File: test/test_updst.cxx (hunk 2 above)
Comment (rcarson3) [reply test/test_updst.cxx]: So, the issue was in our debug builds in which things would break (at least with `rocm v4.x`) as the HIP compiler would not accept things like the below. So, I had to introduce the constexpr of things to get around that issue. I can take a look again and see if the newer versions of rocm accept the earlier code, and if not I'll look at a cleaner workaround with less duplicated code.

``` c++
   // Compiler would error out as it did not recognize the mmodel->nH term as being a constexpr term
   // So, we had to explicitly move to the lower version of things to get things to work...
   // Also worth noting is that none of our other (host or GPU) compilers complained about this before 
   // printVec<mmodel->nH>(&(hist[ecmech::evptn::iHistLbH]), std::cout);
   ecmech::printVec<nH>(&(hist[ecmech::evptn::iHistLbH]), std::cout);
```

File: miniapp/miniapp_util.h (hunk 6 above)
Comment (rcarson3) [reply miniapp/miniapp_util.h]: The boilerplate for the below should still be fine. I might need to update the copyright to 2022 though as that's when I updated things to a more modern version of what the raja team had been using...

File: miniapp/retrieve_kernels.cxx (hunk 9 above)
Comment (rcarson3) [reply miniapp/retrieve_kernels.cxx]: I do know how to eliminate all of the code duplication using templates as we've got the code for it in SNLS.

I think for the time being it'll be okay to keep it as is just because I'm looking to do a series of refactors on ExaCMech to reduce a lot of boilerplate code and make code maintenance hopefully simpler. As part of those refactors, one set of changes will be to require the raja portability suite, so we can leverage all the new abstractions in SNLS and some of the new ones being added as part of https://github.com/LLNL/SNLS/pull/14. With that refactor all of these forall's we have in the code will be reduced down to a single `snls::forall` which handles all the backends for us.

File: src/ecmech/ECMech_gpu_portability.h (hunk 4 above)
Comment (rcarson3) [reply src/ecmech/ECMech_gpu_portability.h]: Taking a look at this again these #ifndef lines related to the above were redundant as we already had a `#pragma once` at the top. So, I've gone ahead and just removed those old lines of code.

File: miniapp/retrieve_kernels.cxx (hunk 9 above)
Comment (rcarson3) [reply miniapp/retrieve_kernels.cxx]: Actually ended up reducing a lot of these GPU calls down to only a few new lines and removed a ton of the repeated lines of code.

File: test/test_updst.cxx (hunk 2 above)
Comment (rcarson3) [reply test/test_updst.cxx]: ROCm still can't deal with the earlier mentioned case. However, I simplified the logic around this code so it should make things simpler going forward.

**Review by nrbarton — APPROVED** [review]
Thanks for tackling the various items that I flagged.
